### DISCUSSION TITLE
Behavior Data Injector

### DISCUSSION THREAD

**Discussion by arzon94** [alandtse/skyrim_vr_address_library#28] [Ideas]
the latest update for the [best dragon overhaul](https://www.nexusmods.com/skyrimspecialedition/mods/19051) requires behavior data injector. 
I anticipate future animation mods to also require this dependency as it enables addition of animation variables & events via without nemesis patch
[link to mod](https://www.nexusmods.com/skyrimspecialedition/mods/78146)
[source code](https://github.com/max-su-2019/BehaviorDataInjector)

**Comment by arzon94** [comment]
Update: doodlezoid has created a [port for AE](https://www.nexusmods.com/skyrimspecialedition/mods/78159?tab=posts). I was hopeful this would work similar to their other mods built with _CommonLibSSE NG_ but I believe this still needs a small addition to vr address library. When I run vr_address_tools I see:

`Finished scanning 12 files. rel_ids: 0 offsets: 0 results: 1
\src/Hook.h:15	ID: 62640	FLAT: 0x140aedd10			WARNING: VR Address undefined.
Found 1 items
`

I'm willing to assist but am unsure on the approach to manually identifying VR addresses

**Reply by alandtse** [reply]
To reverse the address you'd need to use Ida or Ghidra on both a flatrim and VR exe. You then look up what address in VR matches the flat address. Sometimes the function is identical, but other times the assembly is different so the hook will need to be changed.

**Reply by alandtse** [reply]
I found the address. Get doodlezoid to compile with VR support and it should work.

**Reply by arzon94** [reply]
Awesome, I'll send a message
I got this far with flatrim ghidra 
![image](https://user-images.githubusercontent.com/16659722/210693672-7bd11c31-09e5-4c3d-9cf6-7f8be9e21084.png)
to find the address, do you just look for a matching method signature in skyrimvr?

**Reply by alandtse** [reply]
First, you need to remove the Steam DRM before you try to reverse it using [steamless](https://github.com/atom0s/Steamless/releases/tag/v3.1.0.2). Then you allow Ghidra to analyze the flat and vr exes to figure out functions.

Next, you need to use Ghidra's version analysis to try to identify what functions may correspond between the versions. Ghidra has multiple way to match including signature, data matches, and just bindiff. The address tools has a couple of [analysis csvs](https://github.com/alandtse/skyrim_vr_address_library#analysis-csvs) that can also be checked. After using those tools, you can then verify whether the function is the same across versions and whether it's actually equivalent based on calling/callee functions. I've already done this so I can easily check most functions.
